### PR TITLE
KSQL-1782: First draft of KSQL architecture topic

### PR DESCRIPTION
**Pull request by JimGalasyn**

### CODE REVIEW

File: docs/concepts/ksql-architecture.rst [hunk 1]
```diff
@@ -0,0 +1,368 @@
+.. _ksql-architecture:
+
+KSQL Architecture
+#################
+
+The architecture of KSQL enables you to build real-time streaming applications
+from Kafka topics by using only SQL statements and queries. KSQL is built on
+Kafka Streams, so a KSQL application runs on a  Kafka cluster like any other
```
Comment (miguno) [inline docs/concepts/ksql-architecture.rst]: "so a KSQL application runs on a  Kafka cluster like any other Kafka Streams application."

NOOOOOOO

A key misconception that people have is that KSQL runs inside the Kafka cluster.  It does not!  And KStreams apps do not either.

File: docs/concepts/ksql-architecture.rst [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+KSQL has four main components:
+
+* KSQL engine – processes KSQL statements and queries 
+* REST interface – enables client access to the engine
+* KSQL CLI – console that provides a command-line interface (CLI) to the engine
+* KSQL UI – enables developing KSQL applications in |c3|
+
+KSQL Server comprises the KSQL engine and the REST API. KSQL Server instances
+communicate with the Kafka cluster and scale out with load as necessary.
```
Comment (miguno) [inline docs/concepts/ksql-architecture.rst]: This comes across as auto-scaling, which KSQL doesn't support.

File: docs/concepts/ksql-architecture.rst (hunk 1 above)
Comment (JimGalasyn) [reply docs/concepts/ksql-architecture.rst]: Thanks for the catch! Is it correct to say that KSQL runs on a cluster of KSQL servers? Or maybe: a KSQL application _communicates_ with a Kafka cluster?

File: docs/concepts/ksql-architecture.rst [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+    
+    The KSQL REST server is implemented in the
+    `KsqlRestApplication.java <https://github.com/confluentinc/ksql/blob/master/ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java>`__
+    class.
+
+When you deploy your KSQL application, it runs on KSQL Server instances that
+are independent of one another, are fault-tolerant, and can scale elastically
+with load. For more information, see :ref:`ksql-deployment-modes`.
+
+.. image:: ../img/ksql-server-scale-out.gif
```
Comment (joel-hamill) [inline docs/concepts/ksql-architecture.rst]: This animated image seems to load a bit slow. Does it need to be animated? 

image

File: docs/concepts/ksql-architecture.rst [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+:ref:`ksql_capacity_planning`.
+
+KSQL Query Lifecycle
+********************
+
+To create a streaming application with KSQL, you write KSQL statements and
+queries. Each statement and query has a lifecycle with the following steps:
+
+#. You register a KSQL stream or table from an existing Kafka topic with a DDL
+   statement, like CREATE STREAM <my-stream> WITH <topic-name>.
```
Comment (joel-hamill) [inline docs/concepts/ksql-architecture.rst]: should this KSQL be formatted as command syntax?

File: docs/concepts/ksql-architecture.rst [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+
+    CREATE STREAM authorization_attempts 
+      (card_number VARCHAR, attemptTime BIGINT, ...)
+      WITH (kafka_topic='authorizations', value_format=‘JSON’);
+
+KSQL writes DDL and DML statements to the *command topic*. Each KSQL
+Server reads the statement from the command topic, parsing and analyzing
+it.
+
+.. image:: ../img/ksql-deploy-command-topic.gif
```
Comment (joel-hamill) [inline docs/concepts/ksql-architecture.rst]: this animated graphic is a bit tough to follow. i feel like i'm watching a video on mute - not sure what's going on - or what step i'm on. 

image

File: docs/concepts/ksql-architecture.rst [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+For each new DDL statement, the KSQL engine adds an entry to the metastore.
+
+For example, the metastore entry for the previous CREATE STREAM statement might
+resemble:
+
++-------------------------+----------------------------------------------------------------------------------+
+| Source Name             | Structured Data Source                                                           |
++=========================+==================================================================================+
+| AUTHORIZATION_ATTEMPTS  | [DataSourceType: STREAM],                                                        |
+|                         | [Schema:(card_number VARCHAR, attemptTime BIGINT, attemptRegion VARCHAR, ...)],  |
```
Comment (joel-hamill) [inline docs/concepts/ksql-architecture.rst]: Should the example be command syntax - or does that break the table?

File: docs/concepts/ksql-architecture.rst (hunk 1 above)
Comment (cjmatta) [reply docs/concepts/ksql-architecture.rst]: @gwenshap called them a "flock" of KSQL servers ;)

File: docs/concepts/ksql-architecture.rst [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+
+In headless mode you can:
+
+* Start any number of server nodes
+* Pass a SQL file with KSQL statements to execute: ``bin/ksql-node query-file=foo/bar.sql``
+* Version-control your queries and transformations as code
+* Ensure resource isolation
+* Leave resource management to dedicated systems, like Kubernetes
+
+Embedded Deployment
```
Comment (mikebin) [inline docs/concepts/ksql-architecture.rst]: Is embedded mode now an officially supported execution mode for KSQL?

Comment (miguno) [reply docs/concepts/ksql-architecture.rst]: Good catch. It is not (see https://github.com/confluentinc/ksql/issues/734).  Until we support it officially, I wouldn't mention it in our docs. cc @MichaelDrogalis

Comment (MichaelDrogalis) [reply docs/concepts/ksql-architecture.rst]: +1, hold until it's shipped.

File: docs/concepts/ksql-architecture.rst (hunk 3 above)
Comment (MichaelDrogalis) [reply docs/concepts/ksql-architecture.rst]: I agree with Joel. They loaded slowly for me, and it's a bit much to absorb as they scroll by. Static images would be better IMHO.

**Pull request merged**